kong/runloop/certificate.lua

```diff
@@ -238,12 +238,27 @@ local function find_certificate(sni)
     return nil, err
   end
 
-  for _, sni, err in mlcache.each_bulk_res(res) do
+  for _, sni_new, err in mlcache.each_bulk_res(res) do
     if err then
-      log(ERR, "failed to fetch SNI: ", err)
+      -- we choose to not call typedefs.wildcard_host.custom_validator(sni)
+      -- in the front to reduce the cost in normal flow.
+      -- these error messages are from validate_wildcard_host()
+      local idx = err:find("must not have a port", nil, true) or
+                  err:find("must not be an IP", nil, true) or
+                  err:find("invalid value: ", nil, true) or
+                  err:find("only one wildcard must be specified", nil, true) or
+                  err:find("wildcard must be leftmost or rightmost character",
+                           nil, true)
```

Comment:
Hard-coding like this feels like hacky (often forget to update here if err update is called), can't we just log the err directly?

Reply:
What we want to do here is that if the err is from `validate_wildcard_host()`, i.e. the sni the is not valid, we log it as debug level because it's just an invalid request. Otherwise, we log it as error level.
We can do an additional sni validation at the beginning of this function (which looks more elegant). However, this may impact the performance as the regex operations in `validate_wildcard_host` are very expensive. Therefore we changed to this hacky way which at least doesn't increase the cost in normal flow.

Reply:
Could we change the code like below:
```
local patterns = { ... }
local idx
for i, pat in ipairs(patterns) do
 idx = err:find(pat, nil, true)
  if idx then
    break
  end
end
  if idx then
    ....
  end
```

Reply:
Good style. changed.